### PR TITLE
Configure mypy to install dependencies in CI and update pyproject.toml

### PR DESCRIPTION
**Pull request by daniellepintz**
## What does this PR do?

* Update `code-checks.yml` to install all dependencies - fixes #9751
* Fix some of the 36 type errors revealed
* Update `pyproject.toml` for the files that still have type errors

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: .github/workflows/code-checks.yml [hunk 1]
```diff
@@ -18,4 +18,7 @@ jobs:
         run: |
           grep mypy requirements/test.txt | xargs -0 pip install
```
Comment (carmocca) [inline .github/workflows/code-checks.yml:19]: This is not necessary now

```suggestion
```

File: .github/workflows/code-checks.yml [hunk 2]
```diff
@@ -18,4 +18,7 @@ jobs:
...
           pip list
```
Comment (carmocca) [inline .github/workflows/code-checks.yml:20]: This should go after the installation

File: .github/workflows/code-checks.yml [hunk 3]
```diff
@@ -18,4 +18,7 @@ jobs:
...
+          pip install -r requirements.txt
+          pip install “.[dev, examples]”
```
Comment (carmocca) [inline .github/workflows/code-checks.yml:22]: I think this should be enough, we don't run mypy on examples

```suggestion
          pip install '.[dev]'
```

File: .github/workflows/code-checks.yml [hunk 4]
```diff
@@ -18,4 +18,7 @@ jobs:
...
+          pip install pre-commit
```
Comment (carmocca) [inline .github/workflows/code-checks.yml:23]: why install pre-commit?

File: .github/workflows/code-checks.yml (hunk 1 above)
Comment (daniellepintz) [reply .github/workflows/code-checks.yml:19]: isn't it since we are not installing requirements/test which is where mypy is located?

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by carmocca** [comment]
Good, now the failures are showing in our CI.

Do we want to fix all them here, and merge this as a single PR?
Or disable mypy for these files again and address them one by one?

**Comment by daniellepintz** [comment]
@carmocca I'm working on fixing them here

**Comment by daniellepintz** [comment]
For some reason there is still a discrepancy between the CI and mypy locally: https://www.diffchecker.com/rdpVnrN7

**Comment by daniellepintz** [comment]
There is still a strange discrepancy between CI and running mypy locally: https://www.diffchecker.com/mFIxqa2Y

**Review by justusschock — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Comment by awaelchli** [comment]
@daniellepintz @carmocca I suspect there will always be code paths that will be ignored. We have several conditional imports. So unless the mypy check gets included in all the jobs we run, there will always be a discrepancy.

File: pyproject.toml [hunk 5]
```diff
@@ -36,41 +36,31 @@ disable_error_code = "attr-defined"
...
     "pytorch_lightning.callbacks.early_stopping",
     "pytorch_lightning.callbacks.gpu_stats_monitor",
     "pytorch_lightning.callbacks.gradient_accumulation_scheduler",
-    "pytorch_lightning.callbacks.lr_monitor",
     "pytorch_lightning.callbacks.model_summary",
     "pytorch_lightning.callbacks.progress",
     "pytorch_lightning.callbacks.pruning",
     "pytorch_lightning.callbacks.rich_model_summary",
     "pytorch_lightning.core.optimizer",
-    "pytorch_lightning.lite.*",
```
Comment (awaelchli) [inline pyproject.toml:69]: I will take care of it. Already gone through hell (many learnings) to get this line in here. 
#10743

File: pyproject.toml [hunk 6]
```diff
@@ -36,41 +36,31 @@ disable_error_code = "attr-defined"
...
-    "pytorch_lightning.loops.optimization.*",
+    "pytorch_lightning.loops.optimization.closure.py",
+    "pytorch_lightning.loops.optimization.manual_loop.py",
     "pytorch_lightning.loops.evaluation_loop",
-    "pytorch_lightning.trainer.connectors.checkpoint_connector",
-    "pytorch_lightning.trainer.connectors.logger_connector.*",
+    "pytorch_lightning.trainer.connectors.logger_connector.py",
+    "pytorch_lightning.trainer.connectors.logger_connector.fx_validator.py",
```
Comment (carmocca) [inline pyproject.toml:63]: I'll do the logging pieces and loops.

**Pull request merged**